### PR TITLE
DevDocs: Generate a table for displaying proptypes

### PR DESCRIPTION
**Pull request by withinboredom**
Generates a table of proptypes when viewing a component by itself in the DevDocs.

like

#9919 added a json file to the build that uses [react/react-docgen](https://github.com/reactjs/react-docgen) to parse our components and extract the propTypes for each of them. This happens at build time. A later PR may allow this to run similar to how our CSS files are built while developing.

This PR requires that file directly, and doesn't require a server. This has two benefits:

1. It keeps the code simple
2. It doesn't require a running server (offline mode -- as seen above)

A different PR can move the file somewhere in the client path. I'm not particularly happy about a relative path in this code. However, I feel as though that should be separate and not done in this PR.

The styling of this follows @aduth's state selector PR (#9433) style. I've tagged this for design review. I originally experimented with using flex-box instead of a table, but the complexity of the dom and CSS exploded on me. It would be a good future task, though I imagine most developers would be on a desktop device writing code. I could be wrong 😉 ...

There are some interesting possibilities that could come in a future PR near you:

- Instead of using the component's JSDoc for a description, use the readme for that component.
- @retrofox has a really awesome PR (#8482) that could really bring this to life.
- Show an editable example under in the "Use It" box, similar to @enejb's #8284 or my #9289.
- Build the json file if a component changes, similar to how our CSS is currently built.

This work is based on #9289, #9919. There's a fix in #10343 to get some missing components into the index.

---

View live: https://calypso.live/devdocs/blocks?branch=add/proptypes-table

### CODE REVIEW

**Comment by folletto** [comment]
I think design wise it's good for this iteration. I agree it might require tweaking, but it's already good enough, and we can iterate on that once it's in.

The only frustrating bit is that going back to "All Blocks" loses the scroll position in the list, which makes navigating a few blocks fairly frustrating. Might not be something to be done in this PR specifically, but I think it's important to address that.

**Comment by withinboredom** [comment]
@folletto 
> The only frustrating bit is that going back to "All Blocks" loses the scroll position in the list, which makes navigating a few blocks fairly frustrating

Yes, that is really annoying... hopefully I'll figure out a fix soon because it drives me bananas.

**Comment by ehg** [comment]
I noticed after cherry-picking #10343, that inline comments aren't being displayed for `propTypes` documented with inline. e.g. http://calypso.localhost:3000/devdocs/blocks/theme doesn't display a description, though it documents its propTypes: https://github.com/Automattic/wp-calypso/blob/master/client/components/theme/index.jsx#L22

Any idea how we'd fix this?

File: client/blocks/props-viewer/index.jsx [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+
+	render() {
+		const component = this.state.component;
+
+		/**
+		 * Renders a row in the table
+		 * @param {string} propName The name of the prop
+		 * @return {ReactElement} The row
+		 */
+		const row = ( propName ) => {
```
Comment (ehg) [inline client/blocks/props-viewer/index.jsx]: Could we extract functions that render stuff into components?

File: client/blocks/props-viewer/index.jsx [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+			);
+		};
+
+		/**
+		 * Sort prop names by (required, name)
+		 * @param {string} leftName The prop name of the left side
+		 * @param {string} rightName The prop name of the right side
+		 * @return {number} Which side wins
+		 */
+		const sortProps = ( leftName, rightName ) => {
```
Comment (ehg) [inline client/blocks/props-viewer/index.jsx]: This feels like it should live outside `render()` — it also makes me think that we're creating a new function per render?

**Comment by withinboredom** [comment]
@ehg said:
> inline comments aren't being displayed for propTypes

Looks like it only supports docblocks.

It appears like there was [commit](https://github.com/reactjs/react-docgen/commit/acd1d538d44b12e0f1d71839b63e0c3c9990ab00) in react-docgen near where I'd need to update it to support this as it just strips the comments, if I'm reading it correctly. I don't know how the maintainers feel about it ... so I [opened an issue](https://github.com/reactjs/react-docgen/issues/148). Short of that, I think a codemod might would work... I know I've thought about a codemod to move the readme to the component at some point in the future, but I don't think I've brought it up, just thought about it.

File: client/blocks/props-viewer/index.jsx [hunk 3]
```diff
@@ -0,0 +1,192 @@
+/**
+ * External Dependencies
+ */
+
```
Comment (ehg) [inline client/blocks/props-viewer/index.jsx]: No space needed here.

File: client/blocks/props-viewer/index.jsx [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+		return (
+			<div>
+				{ this.props.example }
+				{ this.renderTable( component ) }
+			</div>
+		);
+	}
+
+	static propTypes = {
```
Comment (ehg) [inline client/blocks/props-viewer/index.jsx]: Any benefits to having these at the bottom, rather than the top of the class?

File: Makefile [hunk 5]
```diff
@@ -164,7 +164,7 @@ build-dll: node_modules
 	@mkdir -p build
 	@CALYPSO_ENV=$(CALYPSO_ENV) $(NODE_BIN)/webpack --display-error-details --config webpack-dll.config.js
 
-build-server: install
+build-server: install server/devdocs/proptypes-index.json
```
Comment (ehg) [inline Makefile]: What environments will this actually run in? Do we actually need to generate a full proptypes index in the test environment?

File: client/blocks/props-viewer/index.jsx [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+			</tr>
+		);
+	};
+
+	/**
+	 * Renders a table if it can
+	 * @param {object} component The component to render for
+	 * @return {ReactComponent|null} The table or nothing
+	 */
+	renderTable = ( component ) => {
```
Comment (ehg) [inline client/blocks/props-viewer/index.jsx]: Why not use a regular, unbound, function here?

File: client/blocks/props-viewer/test/index.jsx [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+ * External dependencies
+ */
+import React from 'react';
+import { expect } from 'chai';
+import { shallow } from 'enzyme';
+import PropsViewer from '../index';
+
+const components = require( '../../../../server/devdocs/proptypes-index.json' );
+
+const findRealComponent = ( slug ) => {
```
Comment (ehg) [inline client/blocks/props-viewer/test/index.jsx]: Why are we duplicating code here?

File: client/blocks/props-viewer/index.jsx [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+		if ( ! component ) {
+			return null; //todo: explain why this table is missing
+		}
+
+		return (
```
Comment (ehg) [inline client/blocks/props-viewer/index.jsx]: This still feels a bit long, and unreadable to me.

File: client/blocks/props-viewer/test/index.jsx [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+			const wrapper = shallow( component );
+			expect( wrapper.childAt( 0 ).matchesElement( <div></div> ) ).to.be.true;
+		} );
+	} );
+
+	context( 'renders a table of propTypes', () => {
+		it( 'can render itself', () => {
+			const example = ( <div>Example goes here</div> );
+			const component = ( <PropsViewer component={ 'props-viewer' } example={ example } /> );
+			const spection = findRealComponent( 'props-viewer' )[ 0 ];
```
Comment (ehg) [inline client/blocks/props-viewer/test/index.jsx]: What's a spection?

File: client/blocks/props-viewer/index.jsx (hunk 4 above)
Comment (withinboredom) [inline client/blocks/props-viewer/index.jsx]: I normally see them outside the class at the bottom of the file. I put them at the bottom so that someone would easily find them out of habit, without searching.

File: client/blocks/props-viewer/index.jsx (hunk 3 above)
Comment (withinboredom) [inline client/blocks/props-viewer/index.jsx]: Thanks!

File: Makefile (hunk 5 above)
Comment (withinboredom) [inline Makefile]: I realized I forgot to commit my tests, and the build failed without that file existing. I'd like to chat about it in slack if you don't mind. I don't like it.

File: client/blocks/props-viewer/index.jsx (hunk 8 above)
Comment (withinboredom) [inline client/blocks/props-viewer/index.jsx]: It does feel a bit long, but it's mostly just markup. Lines 155 - 160 could probably become more readable.

File: client/blocks/props-viewer/test/index.jsx (hunk 7 above)
Comment (withinboredom) [inline client/blocks/props-viewer/test/index.jsx]: My personal strategy with duplicate code is to always duplicate once before refactoring for reuse. I do need to address this.

File: client/blocks/props-viewer/test/index.jsx (hunk 9 above)
Comment (withinboredom) [inline client/blocks/props-viewer/test/index.jsx]: Spection comes from Latin specere: to look. 
It technically isn't a real word, but one I made up ;)

File: Makefile (hunk 5 above)
Comment (ehg) [inline Makefile]: Did we converge on a solution to this? :)

File: client/blocks/props-viewer/test/index.jsx (hunk 9 above)
Comment (ehg) [inline client/blocks/props-viewer/test/index.jsx]: Hah, I guess I'd like something a little more human readable though. Makes me feel less of a _caudex_ :P

File: client/blocks/props-viewer/index.jsx (hunk 4 above)
Comment (ehg) [inline client/blocks/props-viewer/index.jsx]: Yeah, I think we only do that because we need `Component` assigned, before we modify `Component.propTypes`. With `createClass`, we'd put them at the top.

File: Makefile (hunk 5 above)
Comment (withinboredom) [inline Makefile]: Yeah, needs mock and this reverted. Still working on it, taking a bit longer than I anticipated.

**Comment by withinboredom** [comment]
I moved the build of the json file to only be when `make run` or `make dashboard` happens instead of on every build. That would mean it won't show on calypso.live or stage ... so, I'm going to have to find a happy place it should be built at. :grumble: ...

File: Makefile [hunk 10]
```diff
@@ -172,7 +172,7 @@ build: install build-$(CALYPSO_ENV)
 
 build-css: public/style.css public/style-rtl.css public/style-debug.css public/editor.css
 
-build-development: server/devdocs/proptypes-index.json server/devdocs/components-usage-stats.json build-server build-dll $(CLIENT_CONFIG_FILE) server/devdocs/search-index.js build-css
+build-development: server/devdocs/components-usage-stats.json build-server build-dll $(CLIENT_CONFIG_FILE) server/devdocs/search-index.js build-css
 
 build-wpcalypso: server/devdocs/proptypes-index.json server/devdocs/components-usage-stats.json build-server build-dll $(CLIENT_CONFIG_FILE) server/devdocs/search-index.js build-css
```
Comment (withinboredom) [inline Makefile]: Now also builds with wpcalypso

File: Makefile [hunk 11]
```diff
@@ -91,10 +91,10 @@ welcome:
 install: node_modules
 
 # Simply running `make run` will spawn the Node.js server instance.
-run: welcome githooks install build
+run: welcome githooks install build server/devdocs/proptypes-index.json
```
Comment (ehg) [inline Makefile]: Superfluous due to https://github.com/Automattic/wp-calypso/pull/10342/files#diff-b67911656ef5d18c4ae36cb6741b7965R171 ?

File: Makefile [hunk 12]
```diff
@@ -91,10 +91,10 @@ welcome:
...
 	@$(NODE) build/bundle-$(CALYPSO_ENV).js
 
-dashboard: install
+dashboard: install server/devdocs/proptypes-index.json
```
Comment (ehg) [inline Makefile]: Also superfluous :)

File: Makefile (hunk 11 above)
Comment (withinboredom) [inline Makefile]: good catch

File: client/blocks/props-viewer/index.jsx [hunk 13]
```diff
@@ -0,0 +1,208 @@
...
+	return components.filter( ( component ) => {
+		return ( slug === component.slug || singular === component.slug ) && component.includePath.indexOf( 'example' ) < 0;
+	} );
+}
+
+/**
+ * Renders a table of prop-types for auto-documentation
+ */
+class PropsViewer extends PureComponent {
+	static displayName = 'PropsViewer';
```
Comment (withinboredom) [inline client/blocks/props-viewer/index.jsx]: Yes, React will do that. Until #10343 gets merged in, the index-generator can't read a class without `displayName` set. Originally the unit tests read this file, so this can be removed now.

File: client/blocks/props-viewer/test/index.jsx [hunk 14]
```diff
@@ -0,0 +1,139 @@
...
+ */
+import React from 'react';
+import { expect } from 'chai';
+import { shallow } from 'enzyme';
+import useMockery from 'test/helpers/use-mockery';
+
+describe( 'PropsViewer', () => {
+	let PropsViewer;
+	useMockery( ( mockery ) => {
+		mockery.registerMock( '../../../server/devdocs/proptypes-index.json', [
```
Comment (withinboredom) [inline client/blocks/props-viewer/test/index.jsx]: Looking into this now, which will also allow for removing `displayName`.

File: client/blocks/props-viewer/index.jsx [hunk 15]
```diff
@@ -0,0 +1,208 @@
...
+import React, { PropTypes, PureComponent } from 'react';
+
+/**
+ * Internal Dependencies
+ */
+import Card from 'components/card';
+import Gridicon from 'components/gridicon';
+
+// todo: move this file!
+const components = require( '../../../server/devdocs/proptypes-index.json' );
```
Comment (withinboredom) [inline client/blocks/props-viewer/index.jsx]: I don't think so... I tried that and it blew up. Either I did it wrong, or `import` tries executing `json` by default.

**Comment by aduth** [comment]
Is there a reason we only show this for Blocks, not UI Components also?

File: client/blocks/props-viewer/index.jsx (hunk 15 above)
Comment (aduth) [inline client/blocks/props-viewer/index.jsx]: >I don't think so... I tried that and it blew up

How did it blow up? Using `import` on my local branch it seems to work fine for me.

Also, is `server` not part of `NODE_PATH` in this case? If it were, we could simply import from `devdocs/proptypes-index.json`.

**Comment by aduth** [comment]
Why do some blocks not have prop types tables despite the underlying component defining its `propTypes`?

e.g. http://calypso.localhost:3000/devdocs/blocks/comment-button

File: client/blocks/props-viewer/index.jsx [hunk 16]
```diff
@@ -0,0 +1,208 @@
...
+		 */
+		component: PropTypes.string.isRequired,
+
+		/**
+		 * The element to display as an example of this component
+		 */
+		example: PropTypes.element.isRequired
+	}
+
+	constructor( props ) {
```
Comment (aduth) [inline client/blocks/props-viewer/index.jsx]: You can leave out `constructor` to fall back to `super`. Only be needed if you have custom constructor logic.

File: client/blocks/props-viewer/index.jsx [hunk 17]
```diff
@@ -0,0 +1,208 @@
...
+	}
+
+	/**
+	 * Set the state of this component to the first matching slug
+	 * @param {string} slug The slug to search for
+	 */
+	setComponent = ( slug ) => {
+		let component = findRealComponent( slug );
+		if ( component.length > 0 ) {
+			component = component[ 0 ];
```
Comment (aduth) [inline client/blocks/props-viewer/index.jsx]: Instead of applying `Array#filter` to the entire set of components only to pick the first matched, seems an appropriate use case for [`_.find`](https://lodash.com/docs/4.17.4#find) (there's a native [`Array#find`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/find) but we can't use it since we don't polyfill ES2015+ prototype methods).

File: client/blocks/props-viewer/index.jsx [hunk 18]
```diff
@@ -0,0 +1,208 @@
...
+
+// todo: move this file!
+const components = require( '../../../server/devdocs/proptypes-index.json' );
+
+/**
+ * Finds a non-example component in the library of components
+ * @param {string} slug The slug to search for
+ * @return {Array} An array of component matches
+ */
+export function findRealComponent( slug ) {
```
Comment (aduth) [inline client/blocks/props-viewer/index.jsx]: Not sure I have a good suggestion, but this function feels a bit fragile with its "singular" logic and testing against example path. Would be nicer if we had solid conventions around singular vs. plural, or exempted `example` components from the propTypes index.

File: client/blocks/props-viewer/index.jsx [hunk 19]
```diff
@@ -0,0 +1,208 @@
...
+		} else {
+			component = null;
+		}
+
+		this.setState( {
+			component
+		} );
+	};
+
+	componentWillMount() {
```
Comment (aduth) [inline client/blocks/props-viewer/index.jsx]: Looks like it's been removed from the official docs, but this seems close to falling within the anti-pattern of using props in state. Here's an archive of the original text:

https://github.com/vjeux/react-website/blob/master/src/react/content/tips/10-props-in-getInitialState-as-anti-pattern.md

tl;dr: It becomes ugly quick to try to keep state in sync in response to props, when you may just as well component result of `setComponent` at time of render (it's probably not as slow as you might think it to be).

**Comment by withinboredom** [comment]
You may need https://github.com/Automattic/wp-calypso/pull/10343 @aduth to get the fix for components not having a `static displayName` set and to get inline comments.

File: client/blocks/props-viewer/index.jsx [hunk 20]
```diff
@@ -0,0 +1,208 @@
...
+	const singular = slug.slice( 0, -1 );
+	return components.filter( ( component ) => {
+		return ( slug === component.slug || singular === component.slug ) && component.includePath.indexOf( 'example' ) < 0;
+	} );
+}
+
+/**
+ * Renders a table of prop-types for auto-documentation
+ */
+class PropsViewer extends PureComponent {
```
Comment (aduth) [inline client/blocks/props-viewer/index.jsx]: Depending on how this component is rendered, it's unlikely it could take advantage of `PureComponent` (the `example` (`element` type) prop will likely never be explicitly equal to previous value).

Related: https://www.youtube.com/watch?v=KYzlpRvWZ6c&feature=youtu.be&t=1132 (on point of `createElement` references changing, `example` will behave like `children`)

File: client/blocks/props-viewer/index.jsx [hunk 21]
```diff
@@ -0,0 +1,208 @@
...
+	 * @param {string} rightName The prop name of the right side
+	 * @return {number} Which side wins
+	 */
+	sortProps = ( leftName, rightName ) => {
+		const component = this.state.component;
+
+		const left = component.props[ leftName ];
+		const right = component.props[ rightName ];
+		if ( left.required === right.required ) {
+			return ( leftName.toLowerCase() < rightName.toLowerCase() ? -1 : 1 );
```
Comment (aduth) [inline client/blocks/props-viewer/index.jsx]: [`String#localeCompare`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/localeCompare) is probably more appropriate here.

File: client/blocks/props-viewer/index.jsx (hunk 8 above)
Comment (aduth) [inline client/blocks/props-viewer/index.jsx]: Could we split this into a few subcomponents? Ideally it'd be nice if we don't fork render logic into separate methods (often that's a sign it should be split to separate components).

File: client/blocks/props-viewer/index.jsx [hunk 22]
```diff
@@ -0,0 +1,208 @@
...
+	 * @param {object} component The component to render for
+	 * @return {ReactComponent|null} The table or nothing
+	 */
+	renderTable( component ) {
+		if ( ! component ) {
+			return null; //todo: explain why this table is missing
+		}
+
+		return (
+			<Card compact={ true } className="props-viewer__card" tagName="div">
```
Comment (aduth) [inline client/blocks/props-viewer/index.jsx]: `div` is already the default prop value for `tagName`, could be omitted.

File: client/blocks/props-viewer/index.jsx [hunk 23]
```diff
@@ -0,0 +1,208 @@
...
+							<tr>
+								<td>Required</td>
+								<td>Name</td>
+								<td>Type</td>
+								<td>Default</td>
+								<td>Description</td>
+							</tr>
+							</thead>
+							<tbody>
+							{ this.rows( component ) }
```
Comment (aduth) [inline client/blocks/props-viewer/index.jsx]: Seems arbitrary to me how we're mixing static methods and instance methods for controlling render behavior of individual component parts.

File: client/blocks/props-viewer/style.scss [hunk 24]
```diff
@@ -0,0 +1,39 @@
+.props-viewer__card {
+  color: darken($gray, 10%);
```
Comment (aduth) [inline client/blocks/props-viewer/style.scss]: Spaces should be tabs.

File: client/blocks/props-viewer/index.jsx [hunk 25]
```diff
@@ -0,0 +1,208 @@
...
+	 * @return {ReactComponent|null} The table or nothing
+	 */
+	renderTable( component ) {
+		if ( ! component ) {
+			return null; //todo: explain why this table is missing
+		}
+
+		return (
+			<Card compact={ true } className="props-viewer__card" tagName="div">
+				<p className="props-viewer__description" >{ PropsViewer.componentDescription( component ) }</p>
```
Comment (aduth) [inline client/blocks/props-viewer/index.jsx]: Remove space before first `>`

File: client/blocks/props-viewer/index.jsx (hunk 23 above)
Comment (aduth) [inline client/blocks/props-viewer/index.jsx]: Should be tabbed one more.

File: client/blocks/props-viewer/style.scss [hunk 26]
```diff
@@ -0,0 +1,39 @@
...
+}
+
+.props-viewer__usage {
+  margin: 28px -25px -17px;
+}
+
+.props-viewer__card p {
```
Comment (aduth) [inline client/blocks/props-viewer/style.scss]: Since the targeted `p` already has a `className`, we should use that: `.props-viewer__description` (in general good to avoid element selectors in styles, unless the cascade is very intentional)

File: client/devdocs/design/search-collection.jsx [hunk 27]
```diff
@@ -64,7 +65,7 @@ const Collection = ( { children, filter, section = 'design', component } ) => {
 				unique={ !! component }
 				url={ exampleLink }
 			>
-				{ example }
+				{ component ? <PropsViewer component={ component } example={ example } /> : example }
```
Comment (aduth) [inline client/devdocs/design/search-collection.jsx]: Why do we pass `example` to `<PropsViewer />` ? Seems we should just leave it where it was at and have `<PropsViewer />` be intentionally concerned only with... well, the props. Also helps avoid the `PureComponent` issue I mentioned in an earlier comment.

Comment (withinboredom) [inline client/devdocs/design/search-collection.jsx]: This is intentional, such that the `PropsViewer` could change the props to the the `example` component. Maybe I'm doing it wrong? I thought about making it a child, but that felt strange since the props are a property of the component and if the component changes in any way, I'd like the props table to rerender. The table, as it stands in this PR, isn't very smart, but it can be quite smart (and show you the props as the example is rendering them).

Comment (aduth) [inline client/devdocs/design/search-collection.jsx]: >This is intentional, such that the `PropsViewer` could change the props to the the `example` component.

Is that coming in a future pull request? I didn't see anything about it here. Maybe a bit premature on how we'd expect that form component to be rendered.

**Comment by withinboredom** [comment]
Closing because I suck at making time for this.

**Comment by DanReyLop** [comment]
In https://github.com/Automattic/wp-calypso/pull/19064 I first made the `generate-proptypes-index` task a bit faster (from 30s to about 25s) but then removed it entirely because it's not being used. So if we resurrect this PR, we'll need to add it again, and I recommend using the optimized version. For that, re-add [this](https://github.com/Automattic/wp-calypso/blob/cee40bd6035348d2ffe27ac22f7e3f1a6c554918/package.json#L128) and [these](https://github.com/Automattic/wp-calypso/blob/cee40bd6035348d2ffe27ac22f7e3f1a6c554918/package.json#L191-L192) lines to `package.json`, and restore [this file](https://github.com/Automattic/wp-calypso/blob/cee40bd6035348d2ffe27ac22f7e3f1a6c554918/server/devdocs/bin/generate-proptypes-index.js).
